### PR TITLE
[CR]New hit/dodge formula

### PR DESCRIPTION
**Pull request by Coolthulhu**
As discussed on forums.

Hit roll is now calculated roughly as:

```
N( hit_dice * 5, 25 ) - dodge_dice * 5 >= 0
```

Where `N( x, y )` is the normal distribution with `x` mean and `y` standard deviation.

This means that:
- Being a single normal distribution, the results should be easy to predict and balance
- Lack of discrete dice means no rounding is necessary. Changed dodge and hit functions to return floats to avoid rounding.
- The formula is much more linear than old multi dice sum and subtraction. For example, hitting a manhack at ~5 total hit dice with old numbers would happen after about 50 attacks, while with the new one it should be closer to 1 in 8. At the same time, normal distribution means hit chance never really goes down to 0, just gets exponentially less likely.

Additional changes:
- Infinite dodges allowed for everyone, each dodge subtracts 2 dodge "dice"
- Removed bouldering and skating downing chances, replaced them with dodge penalty
- Replaced bouldering instant-miss chance with accuracy penalty

WiP because:
- [x] Needs more balancing
- [x] Needs a test verifying and enforcing some expected chances, such as "0 skill 8/8/8/8 survivor has less than 20% but more than 5% chance of hitting a manhack"
- [x] Needs a test verifying that special attack chances are still sane after dodge rolls stopped being random

### CODE REVIEW

**Comment by mugling** [comment]
> Needs a test verifying and enforcing some expected chances

This is a good idea and will help prevent future regressions

**Comment by Lady-Luck** [comment]
This will compile, right? I could certainly test it.

I'll have to set up a new compiling environment for CDDA but it'd be worth it...

File: src/rng.cpp [hunk 1]
```diff
@@ -88,3 +88,9 @@ double rng_normal( double lo, double hi )
     double val = std::normal_distribution<double>( ( hi + lo ) / 2, range )( eng );
     return std::max( std::min( val, hi ), lo );
 }
+
+double normal_roll( double mean, double stddev )
+{
+    static std::default_random_engine eng;
```
Comment (codemime) [inline src/rng.cpp]: The random engine has to be randomly initialized (sorry for the tautology), otherwise it would return same sequence of numbers all the time.

Comment (mugling) [inline src/rng.cpp]: Preumably `rng_normal` has the same issue

**Comment by Coolthulhu** [comment]
Currently I have the following tests (passing):

Attacker | Defender | Expected chance (roughly)
---|---|---
8/8/8/8, no skills, unarmed | Manhack | 15%
8/8/8/8, 3 all skills, two-by-four | Manhack | 30%
10/10/10/10, 8 all skills, katana | Manhack | 70%
Zombie | 8/8/8/8, no skills, unencumbered | 50%
Zombie | 10/10/10/10, 3 all skills, good cotton armor | 20%
Zombie | 10/10/10/10, 8 all skills, survivor suit | 5%
Manhack | 8/8/8/8, no skills, unencumbered | 80%
Manhack | 10/10/10/10, 3 all skills, good cotton armor | 60%
Manhack | 10/10/10/10, 8 all skills, survivor suit | 30%

**Comment by Coolthulhu** [comment]
Should be ready for review now.

**Comment by mugling** [comment]
Looks good. Can you do some more unit tests for player versus, especially player vs zombie

**Comment by Coolthulhu** [comment]
Added character vs zed test and updated the table above.
Fixed standard_npc bug that broke many NPC tests.

**Comment by codemime** [comment]
`normal_roll()` still isn't random. Is that intentional?

**Comment by mugling** [comment]
I'm not so sure those to-hit odds aren't too generous - 60% for an untrained player with a mediocre weapon? That said this PR looks like progress and a solution with unit tests is a good starting point for further balancing.

Future considerations:
* `to_hit` should be much more significant
* `to_hit` needs balancing (weapons should be fast, with good to_hit or high damage - never all)

**Comment by Coolthulhu** [comment]
> I'm not so sure those to-hit odds aren't too generous - 60% for an untrained player with a mediocre weapon?

Basic zombie-brand zombie has 0 dodge and no size modifiers.
Player has +2 to-hit from dexterity.
It makes sense for the most basic enemy to be hit more often than not.

> to_hit should be much more significant

You mean specific kind of to-hit or tighter distribution?
I went with wide distribution to avoid the single biggest issue old system had - accuracy not mattering at all once it goes even a single point above target's dodge.

**Comment by mugling** [comment]
> to_hit should be much more significant

Apart from when you have very poor skills and are fighting certain (annoying) monsters `to_hit` isn't especially relevant. The fact you get +2 from `DEX` only worsens that.

Overall weapon selection is too bland - it's just a quest to get the biggest cutting or bashing weapon, most of which also have high `to_hit` anyway.

Obviously this PR didn't cause that, but for future consideration

**Comment by Coolthulhu** [comment]
> Overall weapon selection is too bland - it's just a quest to get the biggest cutting or bashing weapon, most of which also have high to_hit anyway.

It's probably because of the balancing guidelines which made it so that all weapons are within +1-+3 range.
Should probably be reworked, though that would require some interesting idea.

**Comment by mugling** [comment]
> Should probably be reworked, though that would require some interesting idea.

I have something in progress

**Comment by Zireael07** [comment]
When will this go in? You're debating details outside of this PR and this is sorely needed...

File: src/npc.cpp [hunk 2]
```diff
@@ -114,8 +114,8 @@ standard_npc::standard_npc( const std::string &name, const std::vector<itype_id>
     int_cur = std::max( s_int, 0 );
     int_max = std::max( s_int, 0 );
 
-    for( auto &e: _skills ) {
-        e.second = std::min( std::max( sk_lvl, 0 ), MAX_SKILL );
+    for( auto &e: Skill::skills ) {
+        set_skill_level( e.ident(), std::max( sk_lvl, 0 ) );
```
Comment (mugling) [inline src/npc.cpp]: Why did you remove the bounds checking for `MAX_SKILL`?

File: src/monster.cpp [hunk 3]
```diff
@@ -1384,55 +1393,51 @@ int monster::stability_roll() const
...
-        if(one_in(type->sk_dodge)) {
-            return 0;
-        }
-    }
-
-    int numdice = get_dodge();
-
-    switch (type->size) {
+    float ret = 0.0f;
+    switch( type->size ) {
```
Comment (mugling) [inline src/monster.cpp]: Definition should probably be moved to `JSON` with default value set on load dependent upon size. Doesn't need to be this PR

File: tests/melee_test.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+TEST_CASE("Character attacking a zombie", "[melee]") {
+    monster zed( mtype_id( "mon_zombie" ) );
+    INFO( "Zombie has get_dodge() == " + std::to_string( zed.get_dodge() ) );
+
+    SECTION("8/8/8/8, no skills, unarmed") {
+        standard_npc dude( "TestCharacter", {}, 0, 8, 8, 8, 8 );
+        float prob = brute_probability( dude, zed, num_iters );
+        INFO( full_attack_details( dude ) );
+        THEN( "The chance to hit is near 60%" ) {
+            REQUIRE( prob > 0.5f );
```
Comment (mugling) [inline tests/melee_test.cpp]: Unit tests are far too generous - currently that is ±16% so values could shift a lot and you could end up with two supposedly equivalent tests ~30% apart and still past that test.

File: src/npc.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/npc.cpp]: It's up to test maintainer to enforce the bounds. `MAX_SKILL` isn't enforced globally, so it shouldn't cause tests to act as if it was.

File: tests/melee_test.cpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        dude.weapon = item( "2x4" );
+        float prob = brute_probability( dude, zed, num_iters );
+        INFO( full_attack_details( dude ) );
+        THEN( "The chance to hit is near 70%" ) {
+            REQUIRE( prob > 0.6f );
+            REQUIRE( prob < 0.8f );
+        }
+    }
+
+    SECTION("10/10/10/10, 8 all skills, katana") {
```
Comment (mugling) [inline tests/melee_test.cpp]: Sure, but we should still aim to define them all in one place. This is of benefit if we write further unit tests and/or expand the stats dumper. Being able to refer to S1, S2 .. S9 etc does help as well

**Comment by Coolthulhu** [comment]
I'll push updated (stricter) tests and should be good to go.

File: tests/melee_test.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        float prob = brute_probability( zed, dude, num_iters );
+        INFO( "Has get_dodge() == " + std::to_string( dude.get_dodge() ) );
+        check_near( prob, 0.2f, 0.05f );
+    }
+
+    SECTION("10/10/10/10, 8 all skills, survivor suit") {
+        standard_npc dude( "TestCharacter", { "survivor_suit" }, 8, 10, 10, 10, 10 );
+        float prob = brute_probability( zed, dude, num_iters );
+        INFO( "Has get_dodge() == " + std::to_string( dude.get_dodge() ) );
+        check_near( prob, 0.025f, 0.0125f );
```
Comment (mugling) [inline tests/melee_test.cpp]: Why the different tolerance?

**Comment by mugling** [comment]
Getting multiple test case failures for this, eg.
```
melee_test.cpp:62: FAILED:
  REQUIRE( prob < high )
with expansion:
  0.038f < 0.0375f
with messages:
  Zombie has get_hit() == 4.000000
  Has get_dodge() == 13.000000
```

File: tests/melee_test.cpp (hunk 6 above)
Comment (Coolthulhu) [inline tests/melee_test.cpp]: When the hit chance is very low and very high, the tolerance can be lower because the variance gets lower.
For example, changing from `to_hit - dodge` from 0 to 1 will result in ~10% change in hit ratio, but changing it from 10 to 11 will only change it by something like 2%.
I'd calculate it from the expected value, but that would lead to "ugly" numbers.

**Comment by Coolthulhu** [comment]
Weird, it passes all tests for me.
I guess it must be the seed then. Will need to increase tolerance for some cases.

**Comment by mugling** [comment]
It needs balancing more (for example it's arguably too generous against manhacks) but that can come in a separate PR now we have the test infrastructure.

Once you can get the unit tests passing is good to merge

**Comment by Coolthulhu** [comment]
OK, increased the number of iterations 10 times and tweaked failing tests.
Now it should only fail if something is implementation-dependent.

**Comment by mugling** [comment]
Works for me

**Comment by illi-kun** [comment]
Astyle regression in this PR (changes in ```src/rng.cpp``` makes astyle unhappy).

**Comment by mugling** [comment]
Sorry I merged that last commit without an astyle check. Bet it's a single space or something similar just to spite us
